### PR TITLE
Null safety migration

### PR DESCRIPTION
**Pull request by andzejsw**
All code now is moved to fit null safety logic. Tested on actual iOS device. Works for me currently with build arg `--no-sound-null-safety` but without it doesn't. Will need help with this one.

### CODE REVIEW

**Comment by andzejsw** [comment]
In closest days will probably take a look and fix not covered cases

**Comment by konsultaner** [comment]
> In closest days will probably take a look and fix not covered cases

That would be awesome, I'm toooo busy atm. I'll be able to support in 2-3 weeks earliest.

**Comment by andzejsw** [comment]
Looks like I will need help for not covered test cases. But no rush. Currently I can use code which currently is here.

File: lib/src/authentication/scram_authentication.dart [hunk 1]
```diff
@@ -117,16 +115,16 @@ class ScramAuthentication extends AbstractAuthentication {
...
         ',i=' +
         challengeExtra.iterations.toString();
-    String cBindName = authExtra['channel_binding'];
-    String cBindData = authExtra['cbind_data'];
+    String? cBindName = authExtra['channel_binding'];
+    String? cBindData = authExtra['cbind_data'];
     var cBindFlag = cBindName == null ? 'n' : 'p=' + cBindName;
     var cBindInput =
-        cBindFlag + ',,' + (cBindData == null ? '' : base64.decode(cBindData));
+        cBindFlag + ',,' + (cBindData == null ? '' : base64.decode(cBindData).join()); // TODO: Test if this work. Added jon to decode, cause in context value used as string
```
Comment (konsultaner) [inline lib/src/authentication/scram_authentication.dart:127]: could you explain this line?

File: lib/src/client.dart [hunk 2]
```diff
@@ -83,15 +81,15 @@ class Client {
 
   void _connect(ClientConnectOptions options) async {
     await transport.open(pingInterval: options.pingInterval);
-    if (transport.isOpen) {
+    if (transport.isOpen == true) {
```
Comment (konsultaner) [inline lib/src/client.dart:84]: why?

File: lib/src/message/interrupt.dart [hunk 3]
```diff
@@ -2,9 +2,9 @@ import 'cancel.dart';
 import 'message_types.dart';
 
 class Interrupt {
-  int id;
+  int id = MessageTypes.CODE_INTERRUPT;
```
Comment (konsultaner) [inline lib/src/message/interrupt.dart:5]: this is set in the constructor

File: lib/src/message/publish.dart [hunk 4]
```diff
@@ -3,38 +3,38 @@ import 'message_types.dart';
...
   int requestId;
-  PublishOptions options;
+  PublishOptions? options;
   String topic;
 
   Publish(this.requestId, this.topic,
       {this.options,
-      List<Object> arguments,
-      Map<String, Object> argumentsKeywords}) {
+      List<Object>? arguments,
```
Comment (konsultaner) [inline lib/src/message/publish.dart:11]: shouldn't that become <dynamic> as well?

File: lib/src/protocol/session.dart [hunk 5]
```diff
@@ -83,21 +83,21 @@ class Session {
...
   static Future<Session> start(String realm, AbstractTransport transport,
-      {String authId,
-      List<AbstractAuthentication> authMethods,
-      Duration reconnect}) async {
+      {String? authId,
+      List<AbstractAuthentication>? authMethods,
+      Duration? reconnect}) async {
     /// The realm object my be null bust must mach the uri pattern if it was
     /// passed
-    assert(realm == null || UriPattern.match(realm));
```
Comment (konsultaner) [inline lib/src/protocol/session.dart:96]: why are you removing the null check?

File: lib/src/serializer/json/serializer.dart [hunk 6]
```diff
@@ -222,12 +221,11 @@ class Serializer extends AbstractSerializer {
       }
     }
     _logger.shout('Could not deserialize the message: ' + jsonMessage);
-    // TODO respond with an error
-    return null;
+    return Error(-1, -1, {}, 'Could not deserialize the message: ' + jsonMessage);
```
Comment (konsultaner) [inline lib/src/serializer/json/serializer.dart:224]: this definitly needs a test

File: lib/src/serializer/msgpack/serializer.dart [hunk 7]
```diff
@@ -65,86 +65,85 @@ class Serializer extends AbstractSerializer {
         details.authrole = message[2]['authrole'] ?? '';
         if (message[2]['authextra'] != null) {
           (message[2]['authextra'] as Map).forEach((key, value) {
-            details.authextra ??= <String, dynamic>{};
```
Comment (konsultaner) [inline lib/src/serializer/msgpack/serializer.dart:68]: same question, why removing it?

File: lib/src/serializer/msgpack/serializer.dart [hunk 8]
```diff
@@ -218,12 +217,11 @@ class Serializer extends AbstractSerializer {
       }
     }
     _logger.shout('Could not deserialize the message: ' + msgPack.toString());
-    // TODO respond with an error
-    return null;
+    return Error(-1, -1, {}, 'Could not deserialize the message: ' + msgPack.toString());
```
Comment (konsultaner) [inline lib/src/serializer/msgpack/serializer.dart:220]: needs a test

File: lib/src/serializer/msgpack/serializer.dart [hunk 9]
```diff
@@ -452,19 +450,17 @@ class Serializer extends AbstractSerializer {
...
         detailsParts['authmethods'] = details.authmethods;
       }
-      if (details.authextra != null) {
-        detailsParts['authextra'] = details.authextra;
-      }
+      detailsParts['authextra'] = details.authextra;
       return msgpack_dart.serialize(detailsParts);
     } else {
-      return null;
+      return Uint8List(0); // TODO: check if need to return null
```
Comment (konsultaner) [inline lib/src/serializer/msgpack/serializer.dart:459]: please make sure this works

File: lib/src/transport/socket/socket_transport.dart [hunk 10]
```diff
@@ -182,17 +181,17 @@ class SocketTransport extends AbstractTransport {
     });
     // TODO set keep alive to true
     //_socket.setOption(RawSocketOption.fromBool(??, SO_KEEPALIVE, true), true)
-    return _socket
+    return _socket!
         .where((List<int> message) {
-          message = Uint8List.fromList(_inboundBuffer + message);
-          if (_negotiateProtocol(message) || !_assertValidMessage(message)) {
+          var msg = Uint8List.fromList(_inboundBuffer + message);
```
Comment (konsultaner) [inline lib/src/transport/socket/socket_transport.dart:186]: why are you killing the override?

File: lib/src/transport/socket/socket_transport.dart [hunk 11]
```diff
@@ -342,20 +343,20 @@ class SocketTransport extends AbstractTransport {
...
             .timeout(timeout ?? Duration(seconds: 5));
         return pong;
       } on TimeoutException {
         if (isOpen) {
           rethrow;
         }
-        _pingCompleter.complete();
-        return null;
+        _pingCompleter!.complete();
+        return Uint8List(0); // TODO: check if needs to return null
```
Comment (konsultaner) [inline lib/src/transport/socket/socket_transport.dart:359]: please check it

File: lib/src/transport/socket/socket_transport.dart [hunk 12]
```diff
@@ -371,15 +372,17 @@ class SocketTransport extends AbstractTransport {
       if (_outboundBuffer.isEmpty) {
         _handshakeCompleter.future.then((aVoid) {
           _send0(_outboundBuffer);
-          _outboundBuffer = null;
+          _outboundBuffer = Uint8List(0); // TODO: check if should be _outboundBuffer = null
```
Comment (konsultaner) [inline lib/src/transport/socket/socket_transport.dart:375]: please  check it

**Review by konsultaner — COMMENTED** [review]
In general i'm not sure if the default values will kill some of the wamp scpecs or functionality.

**Comment by johankitelman** [comment]
hey, folks! Any luck with the migration? Would like to give it a try with this package. Thanks!

**Comment by konsultaner** [comment]
@johankitelman I'm waiting for feedback on my review here. The problem with this PR is that @andzejsw did not just add null safty but also altered other code parts. I don't have time to add null safty support at the moment. There are also some coverage issues to fix. 

Feel free to contribute another null safty PR. I'll merge it if no other issues come up.

**Comment by konsultaner** [comment]
@johankitelman since I haven't gotten any response from @andzejsw anymore. I started my own null safty branch. Maybe you could confirm my changes? https://github.com/konsultaner/connectanum-dart/tree/sound-null-safety

**Pull request closed without merging**